### ISSUE TITLE
[FEATURE] Transition states for ConversationHandler

### ISSUE THREAD

**Issue by edenworky** [python-telegram-bot/python-telegram-bot#2388]
I want to have encapsulated (one file for each) nested conversation handlers, but there's no way to trigger callbacks in a parent conversation (despite being able to trigger a parent state with `map_to_parent`. The nested conversation example only works because the different conversations are not encapsulated from each other, so a nested conversation can just directly call a callback the parent conversation uses.

I think it would be great to have transitional states. For an example menu, right now a conversation handler sets up the keyboard in the entry point, then goes to a `MENU` state to handle the updates, and a child conversation needs to call the parent's setup method and then transition to the parent `MENU` (via `map_to_parent`). I want the parent's entry to not reply and instead go to a transition `SETUP_MENU`, which as soon as it's entered sends the menu and transitions to `MENU`. This way, if I want to go back to the parent menu, I can simply transition from the child conversation to the parent `SETUP_MENU` and preserve encapsulation.

I considered that this could possibly be achieved with a state that somehow triggers an update for itself, but I don't know self-trigger like that. Either a special handler or tips for self-triggering would be greatly appreciated, and either way it could simplify the nested conversation example quite substantially imo

**Comment by edenworky** [comment]
I'd also be happy to set up a PR for this given some tips for implementing it

**Comment by Bibo-Joshi** [comment]
Hi.

While I do understand the issue you're facing, I must say that this is just not really how nested conversations work. Just imagine that you have *two* child conversations, that both in the end lead the user back to the main menu, but with a difference in the setup:

- The first child conversation solely uses inline buttons, i.e. the main menu is displayed by calling `update.callback_query.edite_message_text(…)`
- The second child conversation relies on replies from the user, i.e. in the end for the main menu to show you need to send a new message.

Now how would a `SETUP_MENU` "transition state" know which of these to do? In general, your callbacks need to prepare everything the next callback assumes is ready.

If you want to have each child conversation, I recommend to extract logic like "build the main menu" into a standalone file (say `menu_utils.py` or whatever). E.g.:

```python
def main_menu_kwargs(*args, **kwargs) -> Dict[str, Union[str, InlineKeyboardMarkup]]:
 ...
```
Then you can e.g. call `callback_query.edit_message_text(**main_menu_kwargs(…))` *or* `update.effective_message.reply_text(**main_menu_kwargs(…)` depending on what's needed.

**Comment by edenworky** [comment]
Thanks for the quick and detailed reply!

Respectfully, I don't think the problem you raise is valid. States already need to handle this problem all the time. Right now it's solved either by not mixing up states (which could just as well be solved with e.g. `SETUP_INLINE_MENU` and `SETUP_REPLY_MENU`) or making states context-aware, which is also done all the time in normal states. It's not a different kind of problem to managing a menu util that needs to know when to do inline and when not to. 

In summary, you wouldn't always want to hand control over to the parent, but I believe its uses far outweigh the possible confusion resulting from any design choice. To answer your specific question of how `SETUP_MENU` would know what to do, if you don't want a contextual transitional state, don't make one. There are still many valid use cases.

**Comment by Bibo-Joshi** [comment]
> To answer your specific question of how `SETUP_MENU` would know what to do, if you don't want a contextual transitional state, don't make one. There are still many valid use cases.

I don't understand what you're trying to say here. 

>  States already need to handle this problem all the time.

All right, I guess I just didn't understand that you want `SETUP_MENU` be an acutall full blown state with a list of handlers. So I guess the scetch would be something like

```python
ch = ConversationHandler(
    …,
    states: {
        'MENU': 'SETUP_MENU',
        'SETUP_MENU': [MessageHandler(…), CallbackHandler(…)],
    }, …
)
```
and `ConversationHandler.handle_update(…)` would do something like 

```python
new_state = selected_handler.handle_update(…)
if not isinstance(self.states[new_state], list):
    for handler in self.states[new_state]:
        if handler.check_update(update):
            transitioned_to_state = handler.handle_update(…)
            break
...
```

That does sound somewhat reasonable. TBH, though, I'm not too big a fan of adding even more complexity to `ConversationHandler`, especially as this would be a feature exclusively usable for nested conversations. But maybe the other devs feel different about that.

That said, I do think my proposed alternative solution is a viable workorund. The only case where it's not would be if you want to use the same child conversation within different parent conversations, which - sticking to the example - would have differen "main menus". In that case, you could implement a "child factory" along the lines:

```python
def create_child_conversation(main_menu_factory: Callable) -> ConversationHandler:
    def the_callback_that_displays_the_parents_main_menu(update, context):
        # stuff
        update.effective_message.reply_text(**main_menu_factory(...))

    return ConversationHandler( …, states= {…, [SomeHandler(…, callback=the_callback_that_displays_the_parents_main_menu)]})
```

**Comment by edenworky** [comment]
I'll attempt to clarify. Maybe functional state is a better term? I don't want a full-blown state with multiple handlers, I want a state that exists only to set to trigger a callback and move on. This could also be implemented with something like `on_enter`. The sketch would be:

parent: entry -> `SETUP_MENU` (immediately reply with keyboard and continue) -> `MENU` (listen for updates)
And then the child, when ending and returning to the parent, doesn't have to know or care about the keyboard or whatever the parent does next, it can simply map its own `END` state to the parent's `SETUP_MENU`, which would immediately reply and move on to `MENU`.

With a menu factory the child would need to care about what happens after it's done, and is an unnecessary duplication. If I want to do another small step every time before sending the keyboard, I'd need to update both of them.

**Comment by Bibo-Joshi** [comment]
> This could also be implemented with something like `on_enter`.

What do you mean with `on_enter`? 

> parent: entry -> SETUP_MENU (immediately reply with keyboard and continue) -> MENU (listen for updates)

That's basically what I scetched, with the difference that in your scetch

a) it's not clear to me if you want the `SETUP_MENU` callback to even have access to the `update`. If not: Then we're back to "how should the callback know which message to edit in case we're in an inline-menu scenario?"
b) if the callback has access to the `update`, you need to do the `if update.callback_query … else …` manually in the callback and one might as well use handlers, then. 

> If I want to do another small step every time before sending the keyboard, I'd need to update both of them.

If you extract everything needed in both parent and child conversation to a `utils` module, you only need to change that.

> With a menu factory the child would need to care about what happens after it's done, and is an unnecessary duplication.

Let me repeat: "That does sound somewhat reasonable. TBH, though, I'm not too big a fan of adding even more complexity to ConversationHandler […]".
Trying to rephares: My point is that if you move all the "duplicated" code to a helper module, you can minize the duplication with rather small effort. I'm not completely opposed to ease usage of nested conversations, but at least for the time being I'm pointing you to viable workarounds ;)

**Comment by edenworky** [comment]
I greatly appreciate the attention and thought on your end.

To answer your questions respectively, by `on_enter` I mean being able to have a callback be triggered when entering a state. My problem could also be solved by putting the callback `SETUP_MENU` uses as an entry callback for `MENU`.

I do want it to have access to the update, which means the actual callbacks don't need to change; and we're in agreement that my proposal doesn't address the `if update.callback_query ... else ...` pattern you raised, as that isn't the problem I'm trying to solve.

As for the limitations of a util file, I was trying to illustrate this isn't something that can be solved with a menu factory. Except in very simple cases, the only way to use a utils file is have one file for the conversation and one for all the callbacks. Not just a keyboard factory, but the entire conversation logic, otherwise you're risking duplication. This is certainly a viable workaround, but seems needlessly complicated and breaks encapsulation.

Of course, as far as complexity in `ConversationHandler`, you are both expert and arbiter. But assuming a child conversation already passes along its update to the parent, this doesn't even necessarily need to change `ConversationHandler` proper. From what I read in the wiki about custom handlers, it might be possible to just make an autotriggering one, or maybe just hook it up to an internal state change. At most, it would mean adding a "state entry callback" (or `on_enter` or whatever) to `ConversationHandler`, which to my outside observation of `ConversationHandler` internals wouldn't be more than reasonably complex given the potential benefits

**Comment by Bibo-Joshi** [comment]
> I greatly appreciate the attention and thought on your end.

NP :)

> To answer your questions respectively, by on_enter I mean being able to have a callback be triggered when entering a state. 

I see. But, I just remembered, that we also need to think of handlers that have `run_async=True`. In that case, the handler doesn't even acutally return the next state, but just a `promise` (at least currently, but let's not build this thread on assumptions about what happens after #2352 :D). That promise is resolved only when the next update comes in - which is too late to run an "on_entry" callback. TBH this seems somewhat like a dealbreaker for a possible implementation to me (at least currently).

> My problem could also be solved by putting the callback SETUP_MENU uses as an entry callback for MENU.

Not quite: when the child returns tha state that's mapped to a parents state, that parents state is not run. You need a new update for that.

>  Except in very simple cases, the only way to use a utils file is have one file for the conversation and one for all the callbacks.

I can't think of a case where this would be necessary. Can you given an example of where you think that wouldn't work?

> From what I read in the wiki about custom handlers, it might be possible to just make an autotriggering one, or maybe just hook it up to an internal state change.

Do I read that correctly as "Maybe I can subclass ConversationHandler to do what I want"? TBH I'd rather recommend using a utils modle and/or factories. Subclassing CH properly probably won't work unless you fiddle with the internals and then you might end up having everything broken in a next release, because we slightly rearrenged the internals.

**Comment by edenworky** [comment]
Hmm I don't see why the promise stuff is a problem, if `on_entry` (handler? hook?) runs before all the other handlers. Naturally it would be triggered before any update. This would add the additional design constraint that `on_entry` callbacks *can't* be async, which would be quite complex. Of course, you're the expert, and if it's mutually exclusive with async handlers I definitely agree with it being a deal breaker.

> Not quite: when the child returns tha state that's mapped to a parents state, that parents state is not run. You need a new update for that.

That's curious. How does this work? I think I may be confused about how updates work. How do you get the Telegram server to send an update for the parent's state?

 > I can't think of a case where this would be necessary. Can you given an example of where you think that wouldn't work?

I'm not sure precisely what you're referring to with your pronouns here, but if you're asking why having all the callbacks in a separate file as opposed to just e.g. a menu factory, it's because the child needs the setup process and not just the specific factory. If, for example, sending a keyboard is one step of three, I'd need the separate factory, but I'd also need a method doing the full setup, and call it in both conversations. In other words, in the specific case of hooking into an existing state, I want to relinquish control to the state, and the only way to do the exact same thing the parent does to set up the state without duplication would be to share the full callback. Otherwise next time I add a setup step I need to update the child too. Thus you can't implement this with a utils file, only a callbacks file. 

Which isn't like a horrible solution or anything, but it's a bit clunky.

> Do I read that correctly as "Maybe I can subclass ConversationHandler to do what I want"? 

I offered it as a last resort, I certainly don't want us to start fiddling with CH internals. I was referring specifically to custom handlers, specifically https://github.com/python-telegram-bot/python-telegram-bot/wiki/Types-Of-Handlers#custom-updates 

It mentions the possibility of going as far as making handlers for 3rd party stuff. I'm not sure how this would work with what you're saying about updates, but I was thinking about subclassing Handler to trigger a self-update, or at a stretch, listening to however CH internally tracks state, then trigger when the appropriate state is entered. Possible sketch, though not necessarily anything like this:

```python
ConversationHandler(
  states={
    MENU: [
      EntryHandler(setup_menu),
      *normal_handlers
    ]
  }
)
```

**Comment by Bibo-Joshi** [comment]
> Hmm I don't see why the promise stuff is a problem, if `on_entry` (handler? hook?) runs before all the other handlers.
 
> That's curious. How does this work? I think I may be confused about how updates work. How do you get the Telegram server to send an update for the parent's state?

ConversationHandler is a [finite state machine](https://en.wikipedia.org/wiki/Finite-state_machine). If a handler doesn't immediately return the next state (i.e. if running asynchronously), the state is not determined, until we check the next time. We can only reasonably do that when the next update comes in. Sticking to the inline menu example that would be the next button press. But that can't happen if the `on_entry` callback wasn't called - as that can only be called when the handler finally returns the next state.

This might change when we go asyncio, but I'm not sure.

> > I can't think of a case where this would be necessary. Can you given an example of where you think that wouldn't work?
> 
> If, for example, sending a keyboard is one step of three, I'd need the separate factory, but I'd also need a method doing the full setup, and call it in both conversations. In other words, in the specific case of hooking into an existing state, I want to relinquish control to the state, and the only way to do the exact same thing the parent does to set up the state without duplication would be to share the full callback. Otherwise next time I add a setup step I need to update the child too. Thus you can't implement this with a utils file, only a callbacks file.

What exactly do you mean by "step" and "setup" here? If it's anything that can't be extracted to a function that only needs the `update` and `context` arguments, I'm inclined to say that there is a design flaw.

> Which isn't like a horrible solution or anything, but it's a bit clunky.

True

> It mentions the possibility of going as far as making handlers for 3rd party stuff. I'm not sure how this would work with what you're saying about updates, but I was thinking about subclassing Handler to trigger a self-update, or at a stretch, listening to however CH internally tracks state, then trigger when the appropriate state is entered.

While this could work (only when the custom update is a subclass of `telegram.Update`, tough, see the[ note here](https://python-telegram-bot.readthedocs.io/en/v13.2/telegram.ext.conversationhandler.html)), I would rule this a black magic hack and bad design ;)

**Comment by edenworky** [comment]
>  But that can't happen if the on_entry callback wasn't called - as that can only be called when the handler finally returns the next state.

Hmm I think I see the confusion. I want `on_enter` to be a function that only needs the update and context, and the return value is ignored. It can't change states, but it can interact with the API (e.g. send menus). Essentially a place for logic to run which sets up custom logic for a specific state but can't actually interact with the FSM.

To my understanding this doesn't matter for async transitions because it can't do any transitions, simply gets called right after a transition and before any updates. In other words, once a handler returns a new state (async-ly or otherwise), `on_enter` gets called as an additional step for transitioning, with whatever update triggered the state change.

To sketch what I want the code to look like, if now it would need to look like this:

```python
def start_parent_menu(update, context):
  menu_keyboard = ...
  update.reply_text('Choose:', reply_markup=menu_keyboard')
  return MENU

parent = ConversationHandler(
  entry_points=[CommandHandler('start', start_parent_menu],
  states={
    PARENT_MENU: menu_handlers
  },
)

def handle_child_end(update, context):
  start_parent_menu(update, context)
  return END

child = ConversationHandler(
  entry_point=[CallbackQueryHandler(QUERY_FROM_PARENT, handle_child_entry)],
  states={
    CHILD_MENU: [CommandHandler('stop', handle_child_end)]
  },
  map_to_parent={ END: PARENT_MENU }
)
```

To extract it I'd need something like this:

```python
# parent_callbacks.py
def send_parent_menu(update, context)
  menu_keyboard = ...
  update.reply_text('Choose:', reply_markup=menu_keyboard')

# parent.py
from .parent_callbacks import *

def start_parent_menu(update, context):
  send_parent_menu(update, context)
  return MENU

parent = ConversationHandler(
  entry_points=[CommandHandler('start', start_parent_menu],
  states={
    PARENT_MENU: menu_handlers
  },
)

# child.py
from .parent_callbacks import *

def handle_child_end(update, context):
  send_parent_menu(update, context)
  return END

child = ConversationHandler(
  entry_point=[CallbackQueryHandler(QUERY_FROM_PARENT, handle_child_entry)],
  states={
    CHILD_MENU: [CommandHandler('stop', handle_child_end)]
  },
  map_to_parent={ END: PARENT_MENU }
)
```

What I want to do is something like this:
```python
# parent.py
from .parent_callbacks import *

def send_parent_menu(update, context)
  menu_keyboard = ...
  update.reply_text('Choose:', reply_markup=menu_keyboard)

def start_parent_menu(update, context):
  return MENU

# not sure precisely how this would be best implemented
# possible like this:
menu = [EntryHandler(send_parent_menu), *menu_handlers]
# maybe even like this:
menu = {
  'on_enter': send_parent_menu,
  handlers: menu_handlers 
}

parent = ConversationHandler(
  entry_points=[CommandHandler('start', start_parent_menu],
  states={
    PARENT_MENU: menu
  },
)

# child.py
# knowledge of send_parent_menu is unnecessary
def handle_child_end(update, context):
  return END

child = ConversationHandler(
  entry_point=[CallbackQueryHandler(QUERY_FROM_PARENT, handle_child_entry)],
  states={
    CHILD_MENU: [CommandHandler('stop', handle_child_end)]
  },
  map_to_parent={ END: PARENT_MENU }
)
```

I don't remember where, but I'm sure I've seen FSM implementations where this is possible. This sort of design encourages encapsulation and separation of concerns between different conversations. Honestly, even somehow hooking up the child conversation to a parent's entry point rather than state would allow this. I'm certainly not advocating for spooky black magic hacks

**Comment by Bibo-Joshi** [comment]
> To my understanding this doesn't matter for async transitions because it can't do any transitions, simply gets called right after a transition and before any updates.

The point is that with `run_async` that transition doesn't happen. Let me try to sketch the flow in the synchronous case with an easy example:

1.1 User sends `/start`.
2.1. without `on_entry`: CH entry point handler catches the updates, sends the menu keyboard and returns `MENU`.
2.2. with `on_entry`: CH entry point handler catches the updates, and returns `MENU`. `MENU.on_entry` sends the menu keyboard.
3. User presses a button, a `CallbackQueryHandler` in the `MENU` state catches it and returns `CH.END`.

Now asynchronously:

1.1 User sends `/start`.
2.1. without `on_entry`: CH entry point handler catches the updates, sends the menu keyboard and returns a `tg.utils.Promise`. Once the promise is finished, we can check the result and see that it's `MENU` - but we can't know when that is and checking every second clearly is not feasible.
2.2. with `on_entry`: CH entry point handler catches the updates, and returns a promise. `MENU.on_entry` can't run as we don't know when the promise is done and the user doesn't see a keyboard - the conversation is stuck.
3.1. User presses a button. The CH checks if the promise is done. If it is, it knows that the new state is `MENU`. A `CallbackQueryHandler` in the `MENU` state catches it and returns the next promise (which will result in `CH.END`). If it's not, the conversation is not in the `MENU` state so the update is not catched at all. The user needs to press the button again.
3.2. The user sends a random message because they don't know what's going on. Only now CH tries to check if the promise is done. if it is, `MENU.on_entry` is called and sends the keyboard. Now the conversation can continue.

Regarding your "extracting" code example: rename `parent_callabcks.py` to `utils.py` and you basically have what I suggested. If you want to avoid importing from `utils` in `child.py`, I suggest to go with the factory pattern, i.e. use something like I detailed here: https://github.com/python-telegram-bot/python-telegram-bot/issues/2388#issuecomment-778623482 . This still leaves the code of `send_parent_menu` within `parent.py`, thus achieving encapsuliation. But how exactly you do it is basically matter of taste …

**Comment by edenworky** [comment]
Ohh I see the problem now with the async stuff and the updates and such. That is quite a pickle. You could still only have an `on_entry` and explain it doesn't work in the docs, but I can see how that's a whole new can of worms.

As for my extracting code example, I modelled it after your suggestion with `utils` but like I said before:

> As for the limitations of a util file, I was trying to illustrate this isn't something that can be solved with a menu factory. Except in very simple cases, the only way to use a utils file is have one file for the conversation and one for all the callbacks. Not just a keyboard factory, but the entire conversation logic, otherwise you're risking duplication.

As in, as soon as the parent wants to e.g. set up the context or whatever *before* sending the keyboard, and this is part of the `PARENT_MENU` setup which want to run again when the child exits, I'd 2 utils in the util file and call them from the child in the same order as the parent, thus creating duplication (because if the parent changes something in that process, we need to change the child too).

Of course there are many valid uses for a util file, including for holding keyboards, but it doesn't allow me to hand over control to the parent. To truly hand over control, I'd need to call whatever callback the parent is calling, and for that I need to extract all relevant callbacks in full (because of course the parent conv imports the child conv, and they both need to import the callback). In truth I wouldn't be extracting `send_keyboard_menu`, I'd be extracting a `start_parent` which does everything except return the new state (provided by the conversations calling it), which even then means tight coupling and some duplication between the conversations, but I'm more okay with that since we already have `map_to_parent`.

My point in all this is I can work around it, but I still don't think the current design allows me to directly implement the kind of behavior I want, though there are workable solutions and I can see that it would complicate async conversations so much a feature like this would be out of scope until something changes. That all being said, my project doesn't particularly need async (at least for now at least for these things) and I'd still personally enjoy a feature like this.

If, async design constraint nonwithstanding, you think this wouldn't take particularly awful black magic to implement, could you give me a few pointers to implement something like this myself? As far as I care about async stuff, `on_entry` can be called not at all or way too late to be useful. I'd be glad to send as a PR when I'm done for prosperity and whatever might change after updating the async stuff.

Thank you very much for all your patience!

**Comment by Bibo-Joshi** [comment]
> I'd 2 utils in the util file and call them from the child in the same order as the parent

Why not combine them into 1 function? But even with two functions, I don't see why this rules out a factory: just pass multiple functions to the factory.

>  for that I need to extract all relevant callbacks in full

Maybe in your use case, but I'm sure that there are multiple examples where you only need to extract part of the callback.

> My point in all this is I can work around it, but I still don't think the current design allows me to directly implement the kind of behavior I want

I think we can agree on that :D

> If, async design constraint nonwithstanding, you think this wouldn't take particularly awful black magic to implement, could you give me a few pointers to implement something like this myself?

You'd need to call `on_entry` within `ConversationHandler.handle_update`. I guess you'd need to copy-paste the code and insert the logic probably below [this line](https://github.com/python-telegram-bot/python-telegram-bot/blob/a34f0b9beece9f503d0c02dd9ac04477438fa68a/telegram/ext/conversationhandler.py#L562). How exactly you store the `on_entry` callbacks is up to you - maybe just add another dict `on_entry` next to `states` on initialization.

> As far as I care about async stuff, on_entry can be called not at all or way too late to be useful. I'd be glad to send as a PR when I'm done for prosperity and whatever might change after updating the async stuff.

TBH having `on_entry` work only for sync sounds like a half-baked solution to me that will lead to a bunch of confused users in the user group.

Looking at the docs of `asynico.Task`, I see there is a [`add_done_callback`](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.add_done_callback) method which one could use to have the `on_entry` callback run when the task is done. Actually, we *could* try something similar with `utils.Promise`. Then again, it might not be worth implementing that before the switch to asyncio.

Giving it more thought, this might acutally be benefitial not only for nested conversation: In normal conversations you might also have the situation where many different states lead to one somehow special state. Depending on the use case, `on_entry` could save some code duplication.

Still, ConverstationHandler is already rather complex and we have viable workarounds that are not too complex, so this should be discussed carefully.

As the important arguments are somewhat spread across our rather lengthy converstation I think I'll open a new issue were I'll try  to summarize these. We do have some things on our current roadmap though, so I can't make promises how quickly we'll get back to this …

**Comment by Bibo-Joshi** [comment]
Closing in favor of #2390
